test/GPv2Encoding.test.ts

```diff
@@ -369,53 +383,64 @@ describe("GPv2Encoding", () => {
+      const nextAddress = () => ethers.utils.getAddress(nextBytes(20));
+      const nextUint = () => BigNumber.from(nextBytes(32));
+      const interactions = [
+        {
+          target: nextAddress(),
+          value: nextUint(),
+          callData: nextBytes(42),
+        },
+        {
+          target: nextAddress(),
+          value: nextUint(),
+        },
+        {
+          target: nextAddress(),
+          // Note: if the interaction decoding becomes significantly more
+          // inefficient, this test might fail with "tx has a higher gas limit
+          // than the block". In this case, the number of bytes below should be
+          // reduced.
+          callData: nextBytes(12000),
+        },
+        {
+          target: nextAddress(),
+        },
+      ];
 
       const encoder = new SettlementEncoder(testDomain);
-      await encoder.encodeInteraction(interaction1);
-      await encoder.encodeInteraction(interaction2);
-      await encoder.encodeInteraction(interaction3);
+      for (const interaction of interactions) {
+        encoder.encodeInteraction(interaction);
+      }
 
-      const numInteractions = 3;
       const decodedInteractions = await encoding.decodeInteractionsTest(
         encoder.encodedInteractions,
-        numInteractions,
+        interactions.length,
       );
 
-      expect(decodedInteractions.length).to.equal(3);
-      expect(decodedInteractions[0].target).to.equal(interaction1.target);
-      expect(decodedInteractions[0].callData).to.equal(interaction1.callData);
-      expect(decodedInteractions[1].target).to.equal(interaction2.target);
-      expect(decodedInteractions[1].callData).to.equal(interaction2.callData);
-      expect(decodedInteractions[2].target).to.equal(interaction3.target);
-      expect(decodedInteractions[2].callData).to.equal(interaction3.callData);
+      expect(decodedInteractions.length).to.equal(interactions.length);
+      for (let i = 0; i < interactions.length; i++) {
+        const decodedInteraction = decodedInteractions[0];
+        const interaction = interactions[0];
+
+        expect(decodedInteraction.target).to.equal(interaction.target);
+        expect(decodedInteraction.value).to.equal(
+          interaction.value || ethers.constants.Zero,
+        );
+        expect(decodedInteraction.callData).to.equal(
+          interaction.callData || "0x",
+        );
+      }
     });
 
-    // Skipped because this test takes a surprisingly long time to complete.
-    // It is supposed to be passing without any change.
-    it.skip("encoder fails to add interactions with too much calldata", async () => {
+    it("encoder fails to add interactions with too much calldata", async () => {
       const interaction = {
         target: ethers.utils.getAddress(fillBytes(20, 0x00)),
-        callData: fillDistinctBytes(2 ** (8 * 3), 0x00),
+        callData: new Uint8Array(2 ** 24),
       };
```

Comment:
Why is this test so much faster now? The encoding didn't really change. Is it because the bytes are now zero?

Reply:
Its because we avoid a large string concatenation. I imagine that `ethers.utils.solidityPack` simply avoids creating a `2 * 2**24` length string if the `length` argument fails to encode into a `uint24`. `malloc`-ing an unitialized buffer should be very fast.